### ISSUE TITLE
Calling failure(RetrofitError) on 2xx with custom body

### ISSUE THREAD

**Issue by yxuminei** [square/retrofit#786]
Hi,

I decided to move over to Retrofit for my project, so I probably miss something in your documentation. My use case is as follow:

I'm using GSON to parse responses coming from my API Provider, for example, User is defined as the following:

`class User {
    long id;
    String email;
}`

But on certain cases, call to the same api might return

`{ "error": 1234, "message": "User not found"}`

But with httpStatus 200.

If I do

`@GET("/users/{id}")
getDetails(@Path("id"), Callback<User> cb)`

It will still go to `cb.success(User user, Response response)` as the httpStatus is 2xx, but the server decides that it should return the error json described above (for instance, the supplied `{id}` is not a valid id).

Is there a way to do a custom inspection (like checking whether the response json contains the key `error` and `message`) before actually deciding whether to call `cb.success` or `cb.failure` and maybe, do something like `cb.failure(MyCustomError error)`?

**Comment by JakeWharton** [comment]
Replace your use of `Callback<User>` with `MyCallback<User>` (or whatever) that implements this logic and provides abstract methods for what consumers need to implement.
```java
public abstract class MyCallback<T extends BaseResponse> implements Callback<T> {
  public final void success(T body, Response response) {
    if (body.error != null) {
      error(body.message);
    } else {
      success(body);
    }
  }

  public abstract void error(String message);
  public abstract void success(T body);
}
```

**Comment by yxuminei** [comment]
Hi,

Thanks for the quick reply and I got that working :).

However, some APIs expects list of things to be returned e.g. `ArrayList<Action>`, now I have:

`new Callback<ArrayList<Action>>() {...}`, and I can't change it to

`new MyCallback<ArrayList<Action>>() {...}` as `ArrayList` doesn't extend `BaseResponse`.

What would be the best workaround for this case?

Thanks for your time.

**Comment by yxuminei** [comment]
This is what I got so far, is this the recommended / good way of doing this kind of thing?

	private abstract static class ListCallback<T> implements Callback<Response> {
		public final void success(Response body, Response response) {
		}

		public final void failure(RetrofitError retrofitError) {
			if(retrofitError.getKind() == RetrofitError.Kind.HTTP || retrofitError.getKind() == RetrofitError.Kind.CONVERSION) {
				ViewHelper.toast(context, R.string.dayre_exp_difficulty);
			} else {
				ViewHelper.toast(context, R.string.check_network);
			}
		}

		public abstract void done(List<T> list, BaseResponse err);
	}

**Comment by JakeWharton** [comment]
If you do not have a base class how can you have common code which handles errors that are only present in the body?

**Comment by yxuminei** [comment]
@JakeWharton Yup I do have a base class that has `error` and `code`, defined as `BaseResponse`.

The code I wrote in my previous comment is when I am expecting a JSON array coming from the server or JSON object in case of error. I have another callback when I'm expecting a JSON object from the server.

Anyway I've managed to get it working, the above one won't work as I will be getting an exception:
`cannot cast from LinkedTreeMap to CustomObject`

So what I did was to give the class a constructor that accepts my Type:

	Type mType;

	public ListCallback(Type type) {
		mType = type;
	}

which later I do

    ArrayList<T> list = gson.fromJson(result, mType);

All's good now, thanks for your direction, one last thing tho, is there a convenient method to convert from `Response` to `String` instead of doing a `StringBuilder` above?
